### PR TITLE
Feature lesson9

### PR DESCRIPTION
**Pull request by TatsianaKazachenka**

### CODE REVIEW

File: src/test/java/pages/ProductsPage.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+     */
+    public String getProductCountToBadge(){
+        return driver.findElement(PRODUCT_COUNT_BADGE).getText().trim();
+    }
+
+    /**
+     * get all products
+     * @return
+     */
+    public List<WebElement> getElements(){
```
Comment (vchesnakoutest) [inline src/test/java/pages/ProductsPage.java]: getProductElements

File: src/test/java/tests/ProductsTest.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+import tests.Constants.ProductsTestConstants;
+
+import java.lang.reflect.Array;
+import java.util.*;
+
+public class ProductsTest extends BaseTest {
+    @Test
+    public void addProductToCartTest() {
+        addProductToCart();
+
```
Comment (vchesnakoutest) [inline src/test/java/tests/ProductsTest.java:18]: remove space

File: src/test/java/tests/ProductsTest.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    public void removeProductFromCartTest() {
+        addProductToCart();
+
+        cartPage.openPage();
+        boolean isCheckedProduct = cartPage.isCheckedProduct(ProductsTestConstants.PRODUCT_NAME);
+        if (isCheckedProduct) {
+            driver.navigate().back();
+            productsPage.clickProductItem(ProductsTestConstants.PRODUCT_NAME, ProductsTestConstants.REMOVE_PRODUCT_CLASS);
+        }
+        cartPage.openPage();
```
Comment (vchesnakoutest) [inline src/test/java/tests/ProductsTest.java]: that would be enough to remove just one product from Cart in test.
what if isCheckedProduct return false and you'll just opening cart page 2 times ?

File: src/test/java/pages/CartPage.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
+package pages;
+
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import pages.Constants.CartPageConstants;
+
+import java.util.List;
+
+public class CartPage extends BasePage{
```
Comment (vchesnakoutest) [inline src/test/java/pages/CartPage.java]: format code using ctrl+alt+L (need space between BasePage { )

File: src/test/java/pages/Constants/CartPageConstants.java [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package pages.Constants;
+
+public class CartPageConstants {
+    public static final String URL = "https://www.saucedemo.com/cart.html";
```
Comment (vchesnakoutest) [inline src/test/java/pages/Constants/CartPageConstants.java]: you can use BASE_URL = "https://www.saucedemo.com"
and CART_URL = "/cart.html"
BASE_URL + CART_URL

File: src/test/java/pages/Constants/LoginPageConstants.java [hunk 6]
```diff
@@ -0,0 +1,7 @@
+package pages.Constants;
+
+public class LoginPageConstants {
+    public static final String USERNAME = "problem_user";
```
Comment (vchesnakoutest) [inline src/test/java/pages/Constants/LoginPageConstants.java]: better to move test constants with user credentials in Test classes

File: src/test/java/tests/Constants/ProductsTestConstants.java [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package tests.Constants;
+
+public class ProductsTestConstants {
+    public static final String PRODUCT_NAME = "Sauce Labs Fleece Jacket";
+    public static final String PRODUCT_NOT_FOUND = "not found";
+    public static final String ADD_PRODUCT_CLASS = "btn_primary";
```
Comment (vchesnakoutest) [inline src/test/java/tests/Constants/ProductsTestConstants.java]: I think it's better to use html tags classes in locators itself. Not define constants for them

File: src/test/java/tests/LoginTests.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+import tests.Constants.LoginTestsConctants;
+
+public class LoginTests extends BaseTest {
+    public static final By ERROR_MESSAGE_EMPTY = By.xpath("//*[@id='login_button_container']//h3[@data-test='error']");
+
+    @Test
+    public void emptyFieldTest() {
+        loginPage.openPage();
+        loginPage.login("", "");
+
```
Comment (vchesnakoutest) [inline src/test/java/tests/LoginTests.java:15]: remove empty line

File: src/test/java/tests/ProductsTest.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    @Test
+    public void addProductToCartTest() {
+        addProductToCart();
+
+        String price_on_product = productsPage.getProductPrice(ProductsTestConstants.PRODUCT_NAME);
+        cartPage.openPage();
+
+        String price_on_cart = cartPage.isCheckedProduct(ProductsTestConstants.PRODUCT_NAME) ?
+                cartPage.getCartItemPrice(ProductsTestConstants.PRODUCT_NAME) : ProductsTestConstants.PRODUCT_NOT_FOUND;
+        Assert.assertEquals(price_on_cart, price_on_product);
```
Comment (vchesnakoutest) [inline src/test/java/tests/ProductsTest.java]: you can move all asserts to the bottom of the test method. Better when they're defined in "one place"

File: src/test/java/tests/ProductsTest.java [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    public void mappingItemProductsToCartTest() {
+        login();
+        int count = initElements().size();
+        for (int i = 0; i < count; i++){
+            List<WebElement> elements = productsPage.getElements();
+            String nameProduct = elements.get(i).getText();
+            productsPage.clickProductItem(nameProduct, ProductsTestConstants.ADD_PRODUCT_CLASS);
+            cartPage.openPage();
+            String nameProductCart = cartPage.getCartItemName();
+            Assert.assertEquals(nameProduct, nameProductCart);
```
Comment (vchesnakoutest) [inline src/test/java/tests/ProductsTest.java:65]: That's not good practice to use Assert in for cycle. If you want to check all products how they're displayed in Cart, you can get text for all the elements from Products page and put it into Collection. And the same when you opening cart > get text for all opened products and put it also in another Collection. Then outside of for cycle you can compare these 2 collections in Assert. 

We'll learn data providers in future lessons. They will do this work for us in a simple way.

You can leave this test 'as is' now. Just advice for the future.

File: src/test/java/tests/ProductsTest.java [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+        ArrayList<String> sortElementsText = elementsText;
+        Collections.sort(sortElementsText, new Comparator<String>() {
+            public int compare(String s1, String s2) {
+                return s1.toString().compareTo(s2.toString());
+            }
+        });
+        Assert.assertEquals(elementsText, sortElementsText);
+    }
+
+    public void addProductToCart(){
```
Comment (vchesnakoutest) [inline src/test/java/tests/ProductsTest.java]: Don't put methods which are not @Test in Test classes. For your purpose you can create another package "steps" with java classes and put them there. We'll also learn about steps in future lessons.

File: src/test/java/tests/ProductsTest.java [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+
+    public void login(){
+        loginPage.openPage();
+        loginPage.login(LoginPageConstants.USERNAME, LoginPageConstants.PASSWORD);
+    }
+
+    public List<WebElement> initElements(){
+        return productsPage.getElements();
+    }
+}
```
Comment (vchesnakoutest) [inline src/test/java/tests/ProductsTest.java:96]: Good work 👍

File: src/test/java/pages/CartPage.java (hunk 4 above)
Comment (TatsianaKazachenka) [reply src/test/java/pages/CartPage.java]: done

File: src/test/java/pages/Constants/CartPageConstants.java (hunk 5 above)
Comment (TatsianaKazachenka) [reply src/test/java/pages/Constants/CartPageConstants.java]: done

File: src/test/java/pages/Constants/LoginPageConstants.java (hunk 6 above)
Comment (TatsianaKazachenka) [reply src/test/java/pages/Constants/LoginPageConstants.java]: done

File: src/test/java/pages/ProductsPage.java (hunk 1 above)
Comment (TatsianaKazachenka) [reply src/test/java/pages/ProductsPage.java]: done

File: src/test/java/tests/Constants/ProductsTestConstants.java (hunk 7 above)
Comment (TatsianaKazachenka) [reply src/test/java/tests/Constants/ProductsTestConstants.java]: done

File: src/test/java/tests/LoginTests.java (hunk 8 above)
Comment (TatsianaKazachenka) [reply src/test/java/tests/LoginTests.java:15]: done

File: src/test/java/tests/ProductsTest.java (hunk 2 above)
Comment (TatsianaKazachenka) [reply src/test/java/tests/ProductsTest.java:18]: done

File: src/test/java/tests/ProductsTest.java (hunk 9 above)
Comment (TatsianaKazachenka) [reply src/test/java/tests/ProductsTest.java]: done

File: src/test/java/tests/ProductsTest.java (hunk 3 above)
Comment (TatsianaKazachenka) [reply src/test/java/tests/ProductsTest.java]: done

File: src/test/java/tests/ProductsTest.java (hunk 11 above)
Comment (TatsianaKazachenka) [reply src/test/java/tests/ProductsTest.java]: done

File: src/test/java/pages/BasePage.java [hunk 13]
```diff
@@ -4,9 +4,11 @@
 
 abstract class BasePage {
     WebDriver driver;
+    String url;
 
-    BasePage(WebDriver driver) {
+    BasePage(WebDriver driver, String url) {
```
Comment (vchesnakoutest) [inline src/test/java/pages/BasePage.java]: Why you decided to pass String url in page object initialization? I don't think it's a good way to make such initialization.

You can create common method in BasePage:
    
public void openPage(String url) {
        driver.get(url);
    }

and just pass necessary constants on each page, for example:

ProductsPage:

    public void openPage() {
        openPage(SAUCE_DEMO_URL + PRODUCTS_URL);
    }

Similar we did on our latest lesson (29.12)

File: src/test/java/pages/Constants/LoginPageConstants.java (hunk 6 above)
Comment (vchesnakoutest) [reply src/test/java/pages/Constants/LoginPageConstants.java]: please move USERNAME and PASSWORD  to Test Constants

File: src/test/java/tests/ProductsTest.java [hunk 14]
```diff
@@ -6,43 +6,50 @@
 import pages.Constants.LoginPageConstants;
 import tests.Constants.ProductsTestConstants;
 
-import java.lang.reflect.Array;
 import java.util.*;
 
 public class ProductsTest extends BaseTest {
+    public static final String ADD_PRODUCT_CLASS = "btn_primary";
+    public static final String REMOVE_PRODUCT_CLASS = "btn_secondary";
+    public static final String PRODUCT_TYPE_SORT = "az";
```
Comment (vchesnakoutest) [inline src/test/java/tests/ProductsTest.java:12]: It's not correct to place constants in Test classes. Here should be only Test methods.

I suggest you not to put html attributes  "btn_primary", "btn_secondary", "az" in constants.

It's better to refactor this method:
    
public void clickProductItem(String productName, String nameClick) {
        getProduct(productName)
                .findElement(By.xpath(String.format(PRODUCT_TO_CART_BTN, nameClick))).click();
    }

**Option 1:** Divide it into 2 separate methods

public void clickAddProductItem(String productName) {
        getProduct(productName).findElement(By.xpath(PRODUCT_ADD_TO_CART_BTN)).click();
    }

public void clickRemoveProductItem(String productName) {
        getProduct(productName).findElement(By.xpath(PRODUCT_REMOVE_FROM_CART_BTN)).click();
    }

    public static final String PRODUCT_ADD_TO_CART_BTN= "//button[contains(@class, 'btn_primary')]";
    public static final String PRODUCT_REMOVE_FROM_CART_BTN= "//button[contains(@class, 'btn_secondary')]";

**Option 2:** Pass boolean value
    public void clickAddOrRemoveProductItem(String productName, boolean isAddToCart) {
        if(isAddToCart) {
            getProduct(productName).findElement(By.xpath(PRODUCT_ADD_TO_CART_BTN)).click();
        } else {
            getProduct(productName).findElement(By.xpath(PRODUCT_REMOVE_FROM_CART_BTN)).click();
        }
    }

File: src/test/java/pages/Constants/LoginPageConstants.java (hunk 6 above)
Comment (TatsianaKazachenka) [reply src/test/java/pages/Constants/LoginPageConstants.java]: done

File: src/test/java/pages/BasePage.java (hunk 13 above)
Comment (TatsianaKazachenka) [reply src/test/java/pages/BasePage.java]: done

File: src/test/java/tests/ProductsTest.java (hunk 14 above)
Comment (TatsianaKazachenka) [reply src/test/java/tests/ProductsTest.java:12]: done

**Pull request merged**